lib/grandpa/warp_sync.go

```diff
@@ -0,0 +1,337 @@
+// then we verify the justifications against the authorities based on the
+// genesis authorities and the scheduled changes.
+// If we are able to verify all the fragments, then the warp proof is valid.
+func (w *WarpSyncProof) verify(
+	setId grandpa.SetID,
+	authorities primitives.AuthorityList,
+	hardForks map[string]SetIdAuthorityList,
+) (*SetIdAuthorityList, error) {
+	currentSetId := setId
+	currentAuthorities := authorities
+
+	for fragmentNumber, proof := range w.Proofs {
+		headerHash := proof.Header.Hash()
+		number := proof.Header.Number
+
+		hardForkKey := fmt.Sprintf("%v-%v", headerHash, number)
+		if fork, ok := hardForks[hardForkKey]; ok {
+			currentSetId = fork.SetID
+			currentAuthorities = fork.AuthorityList
+		} else {
+			err := proof.Justification.Verify(uint64(currentSetId), currentAuthorities)
+			if err != nil {
+				return nil, err
+			}
+
+			if !bytes.Equal(proof.Justification.Target().Hash.Bytes(), headerHash.ToBytes()) {
+				return nil, fmt.Errorf("mismatch between header and justification")
+			}
+
+			scheduledChange, err := findScheduledChange(proof.Header)
+			if err != nil {
+				return nil, fmt.Errorf("finding scheduled change: %w", err)
+			}
+
+			if scheduledChange != nil {
+				auths, err := grandpaAuthoritiesRawToAuthorities(scheduledChange.Auths)
+				if err != nil {
+					return nil, fmt.Errorf("cannot parse GRANPDA raw authorities: %w", err)
+				}
+
+				currentSetId += 1
+				currentAuthorities = auths
+			} else if fragmentNumber != len(w.Proofs)-1 || !w.IsFinished {
+				return nil, fmt.Errorf("Header is missing authority set change digest")
+			}
+		}
+	}
+
+	return &SetIdAuthorityList{currentSetId, currentAuthorities}, nil
+}
+
+type WarpSyncProofProvider struct {
+	blockState   BlockState
+	grandpaState GrandpaState
+	hardForks    map[string]SetIdAuthorityList
+}
+
+func NewWarpSyncProofProvider(blockState BlockState, grandpaState GrandpaState) *WarpSyncProofProvider {
+	return &WarpSyncProofProvider{
+		blockState:   blockState,
+		grandpaState: grandpaState,
+	}
+}
+
+type SetIdAuthorityList struct {
+	grandpa.SetID
+	primitives.AuthorityList
+}
+
+// Generate build a warp sync encoded proof starting from the given block hash
+func (p *WarpSyncProofProvider) Generate(start common.Hash) ([]byte, error) {
+	// Get and traverse all GRANDPA authorities changes from the given block hash
+	beginBlockHeader, err := p.blockState.GetHeader(start)
+	if err != nil {
+		return nil, fmt.Errorf("%w: %w", errMissingStartBlock, err)
+	}
+
+	lastFinalizedBlockHeader, err := p.blockState.GetHighestFinalisedHeader()
+	if err != nil {
+		return nil, fmt.Errorf("getting best block header: %w", err)
+	}
+
+	if beginBlockHeader.Number > lastFinalizedBlockHeader.Number {
+		return nil, errStartBlockNotFinalized
+	}
+
+	authoritySetChanges, err := p.grandpaState.GetAuthoritiesChangesFromBlock(beginBlockHeader.Number)
+	if err != nil {
+		return nil, err
```

Comment:
I would suggest adding context to the returned error so when debugging makes it easy to trace a possible problem.

```suggestion
		return nil, fmt.Errorf("getting auth changes: %w", err)
```